namer/curatorsd/src/main/scala/com/medallia/l5d/curatorsd/common/CuratorSDCommon.scala

```diff
@@ -0,0 +1,111 @@
+import org.apache.curator.x.discovery.{ServiceDiscoveryBuilder, ServiceInstance}
+import org.codehaus.jackson.map.ObjectMapper
+
+object CuratorSDCommon {
+
+  val curatorClientCache = CacheBuilder.newBuilder().build[String, ServiceDiscoveryInfo]
+
+  /**
+   * @param zkConnectStr ZK connection string
+   * @return Service Discovery set of objects which needs to be closed
+   */
+  def createServiceDiscovery(zkConnectStr: String) = {
+    val serviceDiscoveryInfo = curatorClientCache.get(zkConnectStr, new Callable[ServiceDiscoveryInfo] {
+      def call = ServiceDiscoveryInfo(zkConnectStr)
+    })
+    serviceDiscoveryInfo.addReference()
+    serviceDiscoveryInfo
+  }
+
+  def getServiceFullPath(serviceId: String, tenant: Option[String]): String = {
+    tenant.map(t => Joiner.on(".").join(serviceId, tenant)).getOrElse(serviceId)
+  }
+
+}
+
+/** Scala version of JsonInstanceSerializer (supports scala properties) */
+class ScalaJsonInstanceSerializer[T](val targetClass: Class[T]) extends InstanceSerializer[T] {
+
+  private val objectMapper = new ObjectMapper()
+  private val serviceInstanceClass = objectMapper.getTypeFactory.constructType(classOf[ServiceInstance[T]])
+
+  override def deserialize(bytes: Array[Byte]): ServiceInstance[T] = {
+    val rawServiceInstance: ServiceInstance[T] = objectMapper.readValue(bytes, serviceInstanceClass)
+    targetClass.cast(rawServiceInstance.getPayload) // just to verify that it's the correct type
+    rawServiceInstance.asInstanceOf[ServiceInstance[T]]
+  }
+
+  override def serialize(instance: ServiceInstance[T]): Array[Byte] = {
+    val out = new ByteArrayOutputStream()
+    objectMapper.writeValue(out, instance)
+    out.toByteArray
+  }
+}
+
+case class ServiceDiscoveryInfo(zkConnectStr: String) extends RefCounted {
+
+  private val log = Logger(getClass)
+
+  private val DefaultBaseSleepTime = Duration.fromSeconds(1)
+  private val DefaultMaxRetries = 3
+
+  val curatorClient = CuratorFrameworkFactory.builder
+    .connectString(zkConnectStr)
+    .retryPolicy(new ExponentialBackoffRetry(DefaultBaseSleepTime.inMillis.toInt, DefaultMaxRetries))
+    .build
+
+  val serviceDiscovery = ServiceDiscoveryBuilder.builder(classOf[ServiceInstanceInfo])
+    .client(curatorClient)
+    .serializer(new ScalaJsonInstanceSerializer[ServiceInstanceInfo](classOf[ServiceInstanceInfo]))
+    .basePath("")
+    .build()
+
+  curatorClient.start()
+  curatorClient.blockUntilConnected(10, SECONDS)
+
+  serviceDiscovery.start()
+
+  protected override def performClose() = {
+    log.info("Physically closing curator service discovery %s", zkConnectStr)
+    serviceDiscovery.close()
+    curatorClient.close()
+    log.info("Curator service discovery physically closed")
+  }
+}
+
+trait RefCounted extends Closable {
+
+  private val refCount = new AtomicInteger()
+
+  override def close(deadline: Time): Future[Unit] = Future {
+    val refCountValue = refCount.decrementAndGet()
+    if (refCountValue == 0) {
```

Comment:
This might not work if one of the threads closes while another thread adds a reference before the synchronized block is executed.

Since we are closing all the namers together at the end we can instead have an atomicBoolean
isClosed.
The code would simply do the following if
if(!isClosed.getAndSet())  {
everything else in the close method goes in here.
}